### DISCUSSION TITLE
Unambiguously descriptive protocol encoding

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#171] [Ideas]
I think we need a clear encoding format that does not depend on third-party libraries or the program language.

We can use the `SHA256`, `AES` etc.

But the `HashMap` or others should be discarded, Is there a rule to ensure that all encoding formats are the same between different language or version?

**Comment by lurenpluto** [comment]
Not sure what you mean, SHA256 is a hash algorithm, AES is a symmetric encryption algorithm currently used, and encoding (unified into codec in cyfs-base) has no effect with these two methods

Now cyfs-base uses custom RawCodec byte-based encoding, and HashMap has stable encoding and decoding implementations, and currently ts (js) also has corresponding implementations, and there is no problem with cross-language

**Reply by streetycat** [reply]
> Not sure what you mean, SHA256 is a hash algorithm, AES is a symmetric encryption algorithm currently used, and encoding (unified into codec in cyfs-base) has no effect with these two methods
> 
> Now cyfs-base uses custom RawCodec byte-based encoding, and HashMap has stable encoding and decoding implementations, and currently ts (js) also has corresponding implementations, and there is no problem with cross-language

Great, I see it. I means that:

1. We should make it a rule that everyone can follow and learn from, to prevent others from making similar mistakes, and to provide a basis for reviewers.
    * The standuard algorithms as `SHA256` will be allowed.
    * The format depend on fixed version as `protobuf`, the version should be specificed.
    * Others as `HashMap`, should be encoded byte-by-byte.
2. I didn't see any instructions about this, and not sure if some rules is being followed. I'm not mean there is some problem now.

**Comment by lurenpluto** [comment]
About The Hash Algorithm, currently, the default hash algorithm used in the cyfs-stack is SHA256, which is implicitly set and does not support compatibility with multiple hash algorithms:

- Length limitation for ObjectId/ChunkId
The default length is 32 bytes, so the maximum supported Hash length is 256 bits, which means only SHA256 or lower algorithms can be selected.
- The HASH algorithm is implicitly set
There is no reserved method to determine what HASH algorithm an ObjectId/ChunkId uses. In other words, within the cyfs-stack, if you obtain the binary data of a chunk and want to calculate and verify the ChunkID, you can only use the current cyfs-stack's default hash algorithm. If you modify or replace the SHA256 algorithm, the ChunkID you generate will not be able to be verified correctly in other cyfs-stacks, or you will not be able to correctly verify a received Object/Chunk, unless both parties in communication agree on the same algorithm.

This issue is indeed a headache, and there is no good solution for it yet. If you have any suggestions, please feel free to share them.
